### PR TITLE
Add client side tracking for User Exit Survey Link

### PR DESCRIPTION
**Pull request by tavy88**
### Context

Adding client-side tracking for when the user clicks both of the User Exit Survey links. 
Tracking done inside AppInsights

### Screenshots

Screen Shot 2019-07-11 at 15 10 39

Screen Shot 2019-07-12 at 08 54 51

### Ticket

https://dfedigital.atlassian.net/browse/GET-211

### CODE REVIEW

File: app/webpacker/packs/track-events.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+              appInsights.trackEvent($this.dataset.trackedEvent)
+            }
+          }
+        }
+      });
+    }
+  }
+}
+
+module.exports = new TrackEvents();
```
Comment (tavy88) [inline app/webpacker/packs/track-events.js]: I tried making this as generic as possible and as reusable as possible.

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -51,7 +51,7 @@
           <p class="govuk-phase-banner__content">
             <strong class="govuk-tag govuk-phase-banner__content__tag ">beta</strong>
             <span class="govuk-phase-banner__text">
-              This is a new service – your <%= link_to 'feedback', 'https://www.smartsurvey.co.uk/s/get-help-to-retrain/', class: 'govuk-link', target: '_blank'%> will help us to improve it.
+              This is a new service – your <%= link_to 'feedback', 'https://www.smartsurvey.co.uk/s/get-help-to-retrain/', class: 'govuk-link', target: '_blank', data: { tracked__event: 'User Exit Survey Link', type: 'link' }%> will help us to improve it.
```
Comment (smoothcontract) [inline app/views/layouts/application.html.erb]: This has two underscores in the `data` tag for `tracked__event`

File: app/webpacker/packs/track-events.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+function TrackEvents () {
+  this.start = function () {
+    if (appInsights) {
```
Comment (smoothcontract) [inline app/webpacker/packs/track-events.js]: I don't think this is sufficient to check that `appInsights` variable is defined. Tried this in a browser console on a site without the tracking script loaded (here on github actually) and this seems to raise an error:

```
if (appInsights) { console.log('yes') };
VM373:1 Uncaught ReferenceError: appInsights is not defined
    at <anonymous>:1:1
```

Take a look at this - better to use `typeof` operator I think. https://stackoverflow.com/questions/5113374/javascript-check-if-variable-exists-is-defined-initialized

File: app/webpacker/packs/track-events.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      document.querySelectorAll('[data-tracked-event]').forEach(function (element) {
+        $this = element;
+
+        if ($this.dataset.type === 'link') {
+          $this.onclick = function sendEvent(e) {
+            if ($this.dataset.trackedEvent) {
+              appInsights.trackEvent($this.dataset.trackedEvent)
```
Comment (smoothcontract) [inline app/webpacker/packs/track-events.js]: Is this a synchronous call? What happens if the call takes a while and the link seems unresponsive so user clicks again?

File: app/webpacker/packs/track-events.js (hunk 1 above)
Comment (smoothcontract) [reply app/webpacker/packs/track-events.js]: Cool 👍

File: app/webpacker/packs/track-events.js (hunk 4 above)
Comment (tavy88) [reply app/webpacker/packs/track-events.js]: Fair point.

File: app/webpacker/packs/track-events.js (hunk 3 above)
Comment (tavy88) [reply app/webpacker/packs/track-events.js]: True

File: app/webpacker/packs/track-events.js (hunk 4 above)
Comment (tavy88) [reply app/webpacker/packs/track-events.js]: However, adding this:

```ruby
function TrackEvents () {
  this.start = function () {
    if (typeof(appInsights) === 'object') {
      document.querySelectorAll('[data-tracked-event]').forEach(function (element) {
        $this = element;

        if ($this.dataset.type === 'link') {
          $this.onclick = function sendEvent(e) {
            if ($this.dataset.trackedEvent) {
              appInsights.trackEvent($this.dataset.trackedEvent)
              
              setTimeout(alertFunc, 6000);
            }
          }
        }
      });
    }
  }

  function alertFunc() {
    alert("Hello!");
  }
}
```

Does not prevent the redirect nor stops the page. Even if something breaks in the script, you get an error in the console but you still get redirected to the survey.

Comment (tavy88) [reply app/webpacker/packs/track-events.js]: However, if the code fails it still redirects and the error can be seen in the console.

Comment (tavy88) [reply app/webpacker/packs/track-events.js]: To be safer I decided to use one of JS's native async function `setTimeout`. I initially played with Promises but turns out sync code & promises doesn't play nice. Here's an example of the initial Promise implementation:

```ruby
function trackEvent(eventLabel) {
    const sendEvent = new Promise(function (resolve, reject) {
      try {
        var i = 0;
        while(i < 20) {
          console.log('alert');
          i = i + 1;
        }
        appInsights.trackEvent(eventLabel);
        resolve();
      }
      catch (err) {
        reject(err);
      }
    });
    
    sendEvent
      .then(function success(response) {
        return response;
      })
      .catch(function failure(err) {
        console.error(err);
      });
    }
```

Everything is fine until I add the following block: 

```
var i = 0;
        while(i < 20) {
          console.log('alert');
          i = i + 1;
        }
```
which blocks the main thread, hence my decision to use setTimeout - a native async function.

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (tavy88) [reply app/views/layouts/application.html.erb]: Oops...Copy paste typo.

File: app/webpacker/packs/track-events.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+              }, 0);
+            }
+          };
+        }
+      });
+    }
+  };
+}
+
+export default new TrackEvents();
```
Comment (tavy88) [inline app/webpacker/packs/track-events.js]: To be safer I decided to use one of JS's native async function setTimeout. I initially played with Promises but turns out sync code & promises doesn't play nice. Here's an example of the initial Promise implementation:

function trackEvent(eventLabel) {
    const sendEvent = new Promise(function (resolve, reject) {
      try {
        var i = 0;
        while(i < 20000) {
          console.log('alert');
          i = i + 1;
        }
        appInsights.trackEvent(eventLabel);
        resolve();
      }
      catch (err) {
        reject(err);
      }
    });
    
    sendEvent
      .then(function success(response) {
        return response;
      })
      .catch(function failure(err) {
        console.error(err);
      });
    }
Everything is fine until I add the following block:

var i = 0;
        while(i < 20000) {
          console.log('alert');
          i = i + 1;
        }
which blocks the main thread, hence my decision to use setTimeout - a native async function.

**Comment by cwrw** [comment]
I've had a look through the code, and was hoping to see something use made of javascript functions, as the code currently stands, I found it really hard to follow. Maybe something that looks similar to this might be easier to understand and generic enough as well, what do you think? 
```
function TrackEvents () {
  this.start = function () {
    if(typeof appInsights !== 'undefined') {
      trackLinks();
    }
  }

  function trackLinks() {
    document.querySelectorAll('a[data-tracked-event]').forEach(function(element) {
      element.addEventListener('click', trackInsights, false);
    });
  }

  function trackInsights() {
    if (this && this.hasAttribute('data-type')) {
      appInsights.trackEvent(this.dataset.trackedEvent);
    }
  }
}

module.exports = new TrackEvents();
```

**Comment by cwrw** [comment]
> TrackLinks && TrackInsights is a bit misleading as we do not track 2 different things. We track custom event on links. 
This is an example of how we can structure. Feel free to rename the functions

> Also, your code runs the reporting to App insights sync.
You can add the `setTimeout` inside one of the functions as well

> Also why is typeof appInsights !== 'undefined' better than typeof appInsights === 'object' ?
When having a look through
https://stackoverflow.com/questions/5113374/javascript-check-if-variable-exists-is-defined-initialized
says that However  typeof null will return "object" so we need to actually check for undefined as well as null and not just object

I think generally this is a guidance to make the code more readable than it currently stands and split one function into something procedural to follow

**Comment by tavy88** [comment]
Yeah, I'm not against breaking the code into more functions -- will try to sexy it up a bit.

File: app/webpacker/packs/track-events.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  };
+
+  function sendEvent() {
+    setTimeout(function () {
+      appInsights.trackEvent(this.dataset.trackedEvent);
+    }, 0);
+  };
+}
+
+export default new TrackEvents();
```
Comment (cwrw) [inline app/webpacker/packs/track-events.js]: Is there a difference between `export default new TrackEvents();` and `module.exports = new TrackEvents();` ? If there isn't could we stick to `module.exports = new TrackEvents();` to maintain consistency. see https://github.com/DFE-Digital/get-help-to-retrain/blob/master/app/webpacker/packs/search-form.js#L21

File: app/webpacker/packs/track-events.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+function TrackEvents () {
+  this.start = function () {
+    if (appInsights !== null && typeof(appInsights) !== 'undefined') {
+      document.querySelectorAll('[data-tracked-event]').forEach(function (element) {
```
Comment (cwrw) [inline app/webpacker/packs/track-events.js]: could we also move this to a method? As `this.start` is more of a constructor or initialiser for the application, and ideally will be as  clean as possible, and if we were to extend it with more events to track it would be easier, as so: 
```
 this.start = function () {
    if (appInsights !== null && typeof(appInsights) !== 'undefined') {
       track_clicks()
       track_something_else()
     }
  };
```
It will also help with the readability

File: app/webpacker/packs/track-events.js (hunk 6 above)
Comment (tavy88) [reply app/webpacker/packs/track-events.js]: It's the more modern way of doing it with ES6

File: app/webpacker/packs/track-events.js (hunk 7 above)
Comment (tavy88) [reply app/webpacker/packs/track-events.js]: document.querySelectorAll('[data-tracked-event]') is not tracking just links but tracks any component link, button etc that has this data attributes on. I'd like to keep it like this for now.

File: app/webpacker/packs/track-events.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  this.start = function () {
+    if (typeof(appInsights) !== 'undefined') {
+      document.querySelectorAll('[data-tracked-event]').forEach(function (element) {
+        trackComponentOnClick(element);
+      });
+    }
+  };
+
+  function sendEvent() {
+    var eventLabel = this.dataset.trackedEvent
```
Comment (cwrw) [inline app/webpacker/packs/track-events.js]: 🦂 `;`

Comment (tavy88) [reply app/webpacker/packs/track-events.js]: Ha ha fair...
